Flaky: TestGameServerRestartBeforeReadyCrash

**What happened**:

Occasionally we're seeing the TestGameServerRestartBeforeReadyCrash e2e test fail. Not 100% sure if it's a flaky test, or a bug in code (possible it could be a bug).

**What you expected to happen**:

It passes every time.

**How to reproduce it (as minimally and precisely as possible)**:

Hard to repro, only happens sometimes.

**Anything else we need to know?**:

<details>
<summary>e2e test log:</summary>
<code>
Step #21: time="2019-11-05 14:34:18.378" level=info msg="Waiting for us to have an address to send things to" test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:34:28.776" level=info msg="GameServer created" gs=udp-serverc77fb test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:34:28.777" level=info msg="Dialing UDP message to address" address="35.203.183.113:7847" test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:34:28.777" level=info msg="crashing, and waiting to see restart" test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:34:37.885" level=info msg="sending message" fields.msg=CRASH gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:34:50.147" level=info msg="sending message" fields.msg=CRASH gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:01.498" level=info msg="sending message" fields.msg=CRASH gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:09.559" level=info msg="sending message" fields.msg=CRASH gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:16.957" level=info msg="sending message" fields.msg=CRASH gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:26.471" level=info msg="successfully crashed. Moving on!" test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:43.585" level=info msg="marking GameServer as ready" test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:47.960" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Scheduled test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:35:55.283" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:00.523" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:05.746" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:08.464" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:10.194" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:13.725" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:16.508" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:19.695" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:22.501" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:25.894" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:28.909" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:30.894" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:33.693" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:36.530" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:38.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:42.094" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:44.999" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:48.113" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:51.035" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:53.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:56.972" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:36:59.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:02.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:05.972" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:08.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:11.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:14.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:17.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:20.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:23.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:26.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:29.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:32.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:35.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:38.972" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:41.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:44.971" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:47.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:50.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:53.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:56.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:37:59.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:02.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:05.980" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:08.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:11.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:14.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:17.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:20.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:23.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:26.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:29.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:32.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:35.969" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:38.971" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:41.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:44.971" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:47.970" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: time="2019-11-05 14:38:47.979" level=info msg="sending message" fields.msg=READY gs=udp-serverc77fb state=Unhealthy test=TestGameServerRestartBeforeReadyCrash
Step #21: --- FAIL: TestGameServerRestartBeforeReadyCrash (271.91s)
Step #21:         	Test:       	TestGameServerRestartBeforeReadyCrash
</code>
</details>

Looks to be moving to `Unhealthy` before it should. Need to determine reason.

**Environment**:
- Agones version: 1.2.0-development
- Kubernetes version (use `kubectl version`): 
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.10", GitCommit:"e3c134023df5dea457638b614ee17ef234dc34a6", GitTreeState:"clean", BuildDate:"2019-07-08T03:50:59Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.10-gke.15", GitCommit:"7b5157a7c600aa8ee6e2b36b56c478e030d5bfe2", GitTreeState:"clean", BuildDate:"2019-10-07T20:39:12Z", GoVersion:"go1.11.13b4", Compiler:"gc", Platform:"linux/amd64"}
- Cloud provider or hardware configuration: GKE
- Install method (yaml/helm): helm
- Troubleshooting guide log(s):
  Will go looking to see if we can pull up the sidecar and/or controller logs as well, to see what is forcing the health failure.

- Others: N/A

--- a/pkg/gameservers/controller.go
+++ b/pkg/gameservers/controller.go
@@ -786,6 +786,11 @@ func (c *Controller) syncGameServerRequestReadyState(gs *agonesv1.GameServer) (*
 	for _, cs := range pod.Status.ContainerStatuses {
 		if cs.Name == gs.Spec.Container {
 			if _, ok := gs.ObjectMeta.Annotations[agonesv1.GameServerReadyContainerIDAnnotation]; !ok {
+				// check to make sure this container is actually running. If there was a recent crash, the cache may
+				// not yet have the newer, running container.
+				if cs.State.Running == nil {
+					return nil, errors.New("game server container is not currently running, try again")
+				}
 				gsCopy.ObjectMeta.Annotations[agonesv1.GameServerReadyContainerIDAnnotation] = cs.ContainerID
 			}
 			break

--- a/pkg/gameservers/controller_test.go
+++ b/pkg/gameservers/controller_test.go
@@ -919,6 +919,7 @@ func TestControllerApplyGameServerAddressAndPort(t *testing.T) {
 
 func TestControllerSyncGameServerRequestReadyState(t *testing.T) {
 	t.Parallel()
+	nodeName := "node"
 	containerID := "1234"
 
 	t.Run("GameServer with ReadyRequest State", func(t *testing.T) {
@@ -927,10 +928,14 @@ func TestControllerSyncGameServerRequestReadyState(t *testing.T) {
 		gsFixture := &agonesv1.GameServer{ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: "default"},
 			Spec: newSingleContainerSpec(), Status: agonesv1.GameServerStatus{State: agonesv1.GameServerStateRequestReady}}
 		gsFixture.ApplyDefaults()
-		gsFixture.Status.NodeName = "node"
+		gsFixture.Status.NodeName = nodeName
 		pod, err := gsFixture.Pod()
 		pod.Status.ContainerStatuses = []corev1.ContainerStatus{
-			{Name: gsFixture.Spec.Container, ContainerID: containerID},
+			{
+				Name:        gsFixture.Spec.Container,
+				State:       corev1.ContainerState{Running: &corev1.ContainerStateRunning{}},
+				ContainerID: containerID,
+			},
 		}
 		assert.Nil(t, err)
 		gsUpdated := false
@@ -967,7 +972,11 @@ func TestControllerSyncGameServerRequestReadyState(t *testing.T) {
 		assert.Nil(t, err)
 		pod.Spec.NodeName = nodeFixtureName
 		pod.Status.ContainerStatuses = []corev1.ContainerStatus{
-			{Name: gsFixture.Spec.Container, ContainerID: containerID},
+			{
+				Name:        gsFixture.Spec.Container,
+				State:       corev1.ContainerState{Running: &corev1.ContainerStateRunning{}},
+				ContainerID: containerID,
+			},
 		}
 		gsUpdated := false
 
@@ -1011,11 +1020,15 @@ func TestControllerSyncGameServerRequestReadyState(t *testing.T) {
 		gsFixture := &agonesv1.GameServer{ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: "default"},
 			Spec: newSingleContainerSpec(), Status: agonesv1.GameServerStatus{State: agonesv1.GameServerStateRequestReady}}
 		gsFixture.ApplyDefaults()
-		gsFixture.Status.NodeName = "node"
+		gsFixture.Status.NodeName = nodeName
 		gsFixture.Annotations[agonesv1.GameServerReadyContainerIDAnnotation] = "4321"
 		pod, err := gsFixture.Pod()
 		pod.Status.ContainerStatuses = []corev1.ContainerStatus{
-			{Name: gsFixture.Spec.Container, ContainerID: containerID},
+			{
+				Name:        gsFixture.Spec.Container,
+				State:       corev1.ContainerState{Running: &corev1.ContainerStateRunning{}},
+				ContainerID: containerID,
+			},
 		}
 		assert.Nil(t, err)
 		gsUpdated := false
@@ -1051,6 +1064,34 @@ func TestControllerSyncGameServerRequestReadyState(t *testing.T) {
 			})
 		})
 	}
+
+	t.Run("GameServer whose pod is currently not in a running state, so should retry and not update", func(t *testing.T) {
+		c, m := newFakeController()
+
+		gsFixture := &agonesv1.GameServer{ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: "default"},
+			Spec: newSingleContainerSpec(), Status: agonesv1.GameServerStatus{State: agonesv1.GameServerStateRequestReady}}
+		gsFixture.ApplyDefaults()
+		gsFixture.Status.NodeName = nodeName
+		pod, err := gsFixture.Pod()
+		pod.Status.ContainerStatuses = []corev1.ContainerStatus{{Name: gsFixture.Spec.Container}}
+		assert.Nil(t, err)
+		gsUpdated := false
+
+		m.KubeClient.AddReactor("list", "pods", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			return true, &corev1.PodList{Items: []corev1.Pod{*pod}}, nil
+		})
+		m.AgonesClient.AddReactor("update", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			gsUpdated = true
+			return true, nil, nil
+		})
+
+		_, cancel := agtesting.StartInformers(m, c.podSynced)
+		defer cancel()
+
+		_, err = c.syncGameServerRequestReadyState(gsFixture)
+		assert.EqualError(t, err, "game server container is not currently running, try again")
+		assert.False(t, gsUpdated, "GameServer was updated")
+	})
 
 	t.Run("GameServer with non zero deletion datetime", func(t *testing.T) {
 		testWithNonZeroDeletionTimestamp(t, func(c *Controller, fixture *agonesv1.GameServer) (*agonesv1.GameServer, error) {

--- a/pkg/gameservers/health.go
+++ b/pkg/gameservers/health.go
@@ -232,7 +232,7 @@ func (hc *HealthController) skipUnhealthy(gs *agonesv1.GameServer) (bool, error)
 		return hc.failedContainer(pod), nil
 	}
 
-	// finally, we need to check if there a failed container happened after the gameserver was ready or before.
+	// finally, we need to check if the failed container happened after the gameserver was ready or before.
 	for _, cs := range pod.Status.ContainerStatuses {
 		if cs.Name == gs.Spec.Container {
 			if cs.State.Terminated != nil {

--- a/test/e2e/gameserver_test.go
+++ b/test/e2e/gameserver_test.go
@@ -301,7 +301,7 @@ func TestGameServerRestartBeforeReadyCrash(t *testing.T) {
 		return false
 	})
 	if err != nil {
-		assert.FailNowf(t, "Could not make GameServer Ready: %v", err.Error())
+		assert.FailNowf(t, "Could not make GameServer Ready", "reason: %v", err.Error())
 	}
 	// now crash, should be unhealthy, since it's after being Ready
 	logger.Info("crashing again, should be unhealthy")

